stablehlo/reference/Ops.cpp

```diff
@@ -47,5 +47,37 @@ Tensor eval(SineOp op, const Tensor &operand) {
   return result;
 }
 
+namespace {
+
+// Appies permutation `perm` to an array `array` where perm[i] indicates the
+// location where the current array[i] goes.
+void applyInPlacePermutation(std::vector<int64_t> &array,
+                             const std::vector<int64_t> &perm) {
+  size_t swapIdx;
+  for (size_t i = 0; i < perm.size(); i++) {
+    swapIdx = perm[i];
+    while (swapIdx < i) {
+      swapIdx = perm[swapIdx];
+    }
+    std::swap(array[i], array[swapIdx]);
+  }
+}
+
+}  // namespace
+
+Tensor eval(TransposeOp op, const Tensor &operand, DenseIntElementsAttr perm) {
+  Tensor result(operand);
+  result.setType(op.getType());
+
+  // The operation itself does not require any copying but involves swapping
+  // strides.
+  std::vector<int64_t> stride(operand.getStrides());
+  std::vector<int64_t> permutation(perm.getValues<int64_t>().begin(),
+                                   perm.getValues<int64_t>().end());
+  applyInPlacePermutation(stride, permutation);
+  result.setStrides(stride);
+  return result;
+}
```

Comment:
Similarly to #170,  let's define op implementations from first principles, i.e. using only the definitions in the spec and APIs from `Tensor.h` and `Element.h` which implement these definitions. 

From that perspective, relying on exact representation of a tensor, i.e. something that the spec treats as implementation-defined (currently, this is implicit, #193 makes this explicit), doesn't sound right.

I think improving on this is fairly straightforward if we use the iterator that you proposed in https://github.com/openxla/stablehlo/pull/167 (with the modifications that I proposed in the comments), for example:

```
Tensor eval(TransposeOp op, const Tensor &operand) {
  Tensor result(op.getType());
  for (auto resultIt = result.index_begin(); resultIt != result.index_end(); ++resultIt) {
    auto resultIndex = *resultIt;
    auto operandIndex = permute(resultIt, op.permutation());
    result.set(resultIndex, operandIndex);
  }
  return result;
}
```